R/as-ped-cr.R

```diff
@@ -0,0 +1,195 @@
+#' @param output A chracter value: either "data.frame" or "list". 
+#' Will be matched via \code{match.arg}.
+#' @param ... Further arguments passed to the \code{as_ped} function or its 
+#' methods (\code{data.frame} method) and eventually to 
+#' \code{\link[survival]{survSplit}}
+#' @importFrom Formula Formula
+#' @examples
+#' library(mvna)
+#' data("sir.adm", package = "mvna")
+#' sir_adm <- sir.adm[c(1, 36, 85), ]
+#' sir_adm %>% as_ped_cr(Surv(time, status) ~ age + sex, 
+#'                       cut = list(c(0, 5, 20), c(0, 10, 25)))
+#' sir_adm %>% as_ped_cr(Surv(time, status) ~ age + sex, output = "list")
+#' sir_adm$status <- c("death", "discharge", "cens")
+#' sir_adm %>% as_ped_cr(Surv(time, status) ~ age + sex, censor_code = "cens")
+#' sir_adm %>% as_ped_cr(Surv(time, status) ~ age + sex, output = "list", 
+#'                       censor_code = "cens")
+#' @return A data frame class \code{ped} in piece-wise exponential data format
+#' with an additional column indicating which cause is observed (if output == 
+#' "data.frame".) or a named list of single risk data.frames (if output == 
+#' "list).
+#' @export
+#' @author Philipp Kopper
+as_ped_cr <- function(data, formula, cut = NULL, max_time, censor_code = 0L,
+                      output = c("data.frame", "list"), ...) {
+  df <- check_data(data)
+  data <- df[[1]]
+  f_data <- df[[2]]
+  assert_formula(formula)
+  output <- match.arg(output, c("data.frame", "list"))
+  time_str <- all.vars(formula)[1L]
+  status_str <- all.vars(formula)[2L]
+  true_time <- f_data[[time_str]]
+  f_data <- make_numeric(f_data, status_str, censor_code)
+  true_status <- f_data[[status_str]]
+  if (is.data.frame(data)) {
+    status <- unique(data[[status_str]])
+  } else {
+    status <- unique(data[[1L]][[status_str]])
+  }
+  status_num <- unique(f_data[[status_str]])
+  status <- status[status != censor_code]
+  if (length(status) < 2) {
+    stop("There are no competing risks. Use as_ped() instead.")
+  } 
+  cut <- check_cuts(cut, status)
+  ped_sets <- vector(mode = "list", length = length(status))
+  for (i in 1:length(status)) {
+    current_data <- f_data
+    current_data[[status_str]][f_data[[status_str]] != status_num[i]] <- 0L
+    current_data[[status_str]][f_data[[status_str]] == status_num[i]] <- 1L
+    if (!is.data.frame(data) & length(data) == 2L) {
+      current_data <- list(current_data, data[[2L]])
+    }
+    ped_sets[[i]] <- as_ped(data = current_data, formula = formula, 
+                            cut = cut[[i]], ...)
+    if (output != "list") {
+      ped_sets[[i]]$cause <- as.factor(as.character(status[i]))
+    }
+    class(ped_sets[[i]]) <- c("ped", "data.frame")
+  }
+  if (output == "list") {
+    ped <- ped_sets
+    names(ped) <- status
+    class(ped) <- c("ped_cr_list", "ped_cr")
+    attr(ped, "risks") <- status
+    return(ped)
+  } else {
+    ped <- as.data.frame(Reduce("rbind", ped_sets))
```

Comment:
`do.call(rbind, ped_sets)` should be more efficient

Reply:
Cheers